File: sdk/typescript/src/namespaces/email-priority.ts
/**
 * Email Priority Namespace API
 *
 * Provides email prioritization and inbox management:
 * - Email priority scoring with ML/LLM tiers
 * - Inbox ranking and categorization
 * - User feedback for learning
 * - VIP sender management
 * - Gmail OAuth integration
 * - Cross-channel context boosts
 */

/**
 * Priority scoring tier.
 */
export type ScoringTier = 'tier_1_rules' | 'tier_2_ml' | 'tier_3_llm';

/**
 * User action types for feedback.
 */
export type UserAction =
  | 'read'
  | 'archived'
  | 'deleted'
  | 'replied'
  | 'starred'
  | 'important'
  | 'snoozed';

/**
 * Email category.
 */
export type EmailCategory =
  | 'invoices'
  | 'receipts'
  | 'newsletters'
  | 'promotions'
  | 'personal'
  | 'work'
  | 'other';

/**
 * Gmail OAuth scopes.
 */
export type GmailScopes = 'readonly' | 'full';

/**
 * Email message for prioritization.
 */
export interface EmailMessage {
  id: string;
  thread_id?: string;
  subject: string;
  from_address: string;
  to_addresses?: string[];
  cc_addresses?: string[];
  bcc_addresses?: string[];
  date?: string;
  body_text?: string;
  body_html?: string;
  snippet?: string;
  labels?: string[];
  headers?: Record<string, string>;
  is_read?: boolean;
  is_starred?: boolean;
  is_important?: boolean;
  has_attachments?: boolean;
}

/**
 * Priority factor contributing to score.
 */
export interface PriorityFactor {
  name: string;
  weight: number;
  value: number;
  reason: string;
}

/**
 * Priority scoring result.
 */
export interface PriorityResult {
  email_id: string;
  priority_score: number;
  confidence: number;
  tier_used?: ScoringTier;
  rationale: string;
  factors?: PriorityFactor[];
}

/**
 * Category result.
 */
export interface CategoryResult {
  email_id: string;
  category: EmailCategory;
  confidence: number;
  sub_category?: string;
}

/**
 * Category statistics.
 */
export interface CategoryStats {
  total: number;
  by_category: Record<EmailCategory, number>;
  avg_confidence: number;
}

/**
 * Email configuration.
 */
export interface EmailConfig {
  vip_domains: string[];
  vip_addresses: string[];
  internal_domains: string[];
  auto_archive_senders: string[];
  tier_1_confidence_threshold: number;
  tier_2_confidence_threshold: number;
  enable_slack_signals: boolean;
  enable_calendar_signals: boolean;
  enable_drive_signals: boolean;
}

/**
 * Feedback item for batch recording.
 */
export interface FeedbackItem {
  email_id: string;
  action: UserAction;
  email?: EmailMessage;
  response_time_minutes?: number;
}

/**
 * Inbox item with email and priority.
 */
export interface InboxItem {
  email: EmailMessage;
  priority: PriorityResult;
}

/**
 * Inbox fetch parameters.
 */
export interface InboxParams {
  limit?: number;
  offset?: number;
  labels?: string[];
  unread_only?: boolean;
}

/**
 * Cross-channel context.
 */
export interface CrossChannelContext {
  email_address: string;
  slack_activity?: {
    recent_messages: number;
    active_channels: string[];
    last_active: string;
  };
  calendar_events?: {
    upcoming_meetings: number;
    shared_meetings: number;
  };
  drive_activity?: {
    shared_documents: number;
    recent_collaborations: number;
  };
}

/**
 * Context-based priority boost.
 */
export interface ContextBoost {
  email_id: string;
  total_boost: number;
  slack_activity_boost: number;
  drive_relevance_boost: number;
  calendar_urgency_boost: number;
  slack_reason?: string;
  drive_reason?: string;
  calendar_reason?: string;
  related_slack_channels?: string[];
  related_drive_files?: string[];
  related_meetings?: string[];
}

/**
 * Client interface for email operations.
 */
interface EmailPriorityClientInterface {
  request<T = unknown>(
    method: string,
    path: string,
    options?: { params?: Record<string, unknown>; json?: Record<string, unknown> }
  ): Promise<T>;
}

/**
 * Email Priority API namespace.
 *
 * Provides methods for email prioritization and inbox management:
 * - Score and rank emails by priority
 * - Categorize emails into smart folders
 * - Record user feedback for ML learning
 * - Manage VIP senders
 * - Connect Gmail via OAuth
 *
 * @example
 * ```typescript
 * const client = createClient({ baseUrl: 'https://api.aragora.ai' });
 *
 * // Score a single email
 * const result = await client.emailPriority.prioritize(email);
 *
 * // Rank entire inbox
 * const ranked = await client.emailPriority.rankInbox(emails);
 *
 * // Record user action for learning
 * await client.emailPriority.recordFeedback('msg_123', 'replied');
 * ```
 */
export class EmailPriorityAPI {
  constructor(private client: EmailPriorityClientInterface) {}

  // =========================================================================
  // Priority Scoring
  // =========================================================================

  /**
   * Score a single email for priority.
   */
  async prioritize(
    email: EmailMessage,
    forceTier?: ScoringTier
  ): Promise<{ success: boolean; result: PriorityResult }> {
    return this.client.request('POST', '/api/v1/email/prioritize', {
      json: { email, force_tier: forceTier },
    });
  }

  /**
   * Rank multiple emails by priority.
   */
  async rankInbox(
    emails: EmailMessage[],
    limit?: number
  ): Promise<{ success: boolean; results: PriorityResult[]; total: number }> {
    return this.client.request('POST', '/api/v1/email/rank-inbox', {
      json: { emails, limit },
    });
  }

  // =========================================================================
  // Feedback and Learning
  // =========================================================================

  /**
   * Record user action for ML learning.
   */
  async recordFeedback(
    emailId: string,
    action: UserAction,
    email?: EmailMessage
  ): Promise<{ success: boolean }> {
    return this.client.request('POST', '/api/v1/email/feedback', {
      json: { email_id: emailId, action, email },
    });
  }

  /**
   * Record batch of user actions.
   */
  async recordFeedbackBatch(
    items: FeedbackItem[]
  ): Promise<{ recorded: number; errors: number }> {
    return this.client.request('POST', '/api/v1/email/feedback/batch', {
      json: { items },
    });
  }

  // =========================================================================
  // Categorization
  // =========================================================================

  /**
   * Categorize an email to a smart folder.
   */
  async categorize(
    email: EmailMessage
  ): Promise<{ success: boolean; result: CategoryResult }> {
    return this.client.request('POST', '/api/v1/email/categorize', {
      json: { email },
    });
  }

  /**
   * Categorize multiple emails.
   */
  async categorizeBatch(
    emails: EmailMessage[],
    concurrency?: number
  ): Promise<{ results: CategoryResult[]; stats: CategoryStats }> {
    return this.client.request('POST', '/api/v1/email/categorize/batch', {
      json: { emails, concurrency },
    });
  }

  /**
   * Apply Gmail label based on category.
   */
  async applyLabel(
    emailId: string,
    category: EmailCategory
  ): Promise<{ success: boolean; label_applied: string }> {
    return this.client.request('POST', '/api/v1/email/categorize/apply-label', {
      json: { email_id: emailId, category },
    });
  }

  // =========================================================================
  // Inbox Management
  // =========================================================================

  /**
   * Fetch and rank inbox emails.
   */
  async fetchInbox(
    params?: InboxParams
  ): Promise<{ success: boolean; inbox: InboxItem[]; total: number; fetched_at: string }> {
    return this.client.request('GET', '/api/v1/email/inbox', {
      params: params as Record<string, unknown>,
    });
  }

  // =========================================================================
  // Configuration
  // =========================================================================

  /**
   * Get email prioritization configuration.
   */
  async getConfig(): Promise<{ success: boolean; config: EmailConfig }> {
    return this.client.request('GET', '/api/v1/email/config');
  }

  /**
   * Update email prioritization configuration.
   */
  async updateConfig(
    updates: Partial<EmailConfig>
  ): Promise<{ success: boolean; config: EmailConfig }> {
    return this.client.request('PUT', '/api/v1/email/config', {
      json: updates,
    });
  }

  // =========================================================================
  // VIP Management
  // =========================================================================

  /**
   * Add a VIP email address or domain.
   */
  async addVip(options: {
    email?: string;
    domain?: string;
  }): Promise<{ success: boolean; message: string }> {
    return this.client.request('POST', '/api/v1/email/vip', {
      json: options,
    });
  }

  /**
   * Remove a VIP email address or domain.
   */
  async removeVip(options: {
    email?: string;
    domain?: string;
  }): Promise<{ success: boolean; message: string }> {
    return this.client.request('DELETE', '/api/v1/email/vip', {
      json: options,
    });
  }

  // =========================================================================
  // Gmail OAuth
  // =========================================================================

  /**
   * Get Gmail OAuth authorization URL.
   */
  async getGmailOAuthUrl(
    redirectUri: string,
    state?: string,
    scopes?: GmailScopes
  ): Promise<{ success: boolean; oauth_url: string; scopes: GmailScopes }> {
    return this.client.request('POST', '/api/v1/email/gmail/oauth/url', {
      json: { redirect_uri: redirectUri, state, scopes },
    });
  }

  /**
   * Handle Gmail OAuth callback.
   */
  async handleGmailOAuthCallback(
    code: string,
    redirectUri: string
  ): Promise<{ success: boolean; email?: string; message: string }> {
    return this.client.request('POST', '/api/v1/email/gmail/oauth/callback', {
      json: { code, redirect_uri: redirectUri },
    });
  }

  /**
   * Check Gmail connection status.
   */
  async getGmailStatus(): Promise<{
    authenticated: boolean;
    email?: string;
    scopes?: string[];
    expires_at?: string;
  }> {
    return this.client.request('GET', '/api/v1/email/gmail/status');
  }

  // =========================================================================
  // Cross-Channel Context
  // =========================================================================

  /**
   * Get cross-channel context for an email address.
   */
  async getContext(emailAddress: string): Promise<{ context: CrossChannelContext }> {
    return this.client.request('GET', `/api/v1/email/context/${encodeURIComponent(emailAddress)}`);
  }

  /**
   * Get context-based priority boosts for an email.
   */
  async getContextBoost(email: EmailMessage): Promise<{ success: boolean; boost: ContextBoost }> {
    return this.client.request('POST', '/api/v1/email/context/boost', {
      json: { email },
    });
  }
}

File: sdk/python/aragora_sdk/namespaces/email_priority.py
"""
Email Priority Namespace API

Provides email prioritization and inbox management:
- Email priority scoring with ML/LLM tiers
- Inbox ranking and categorization
- User feedback for learning
- VIP sender management
- Gmail OAuth integration

Features:
- Multi-tier priority scoring
- Smart email categorization
- Cross-channel context boosts
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient


ScoringTier = Literal["tier_1_rules", "tier_2_ml", "tier_3_llm"]
UserAction = Literal["read", "archived", "deleted", "replied", "starred", "important", "snoozed"]
EmailCategory = Literal[
    "invoices", "receipts", "newsletters", "promotions", "personal", "work", "other"
]
GmailScopes = Literal["readonly", "full"]


class EmailPriorityAPI:
    """
    Synchronous Email Priority API.

    Provides methods for email prioritization and inbox management:
    - Score and rank emails by priority
    - Categorize emails into smart folders
    - Record user feedback for ML learning
    - Manage VIP senders
    - Connect Gmail via OAuth

    Example:
        >>> client = AragoraClient(base_url="https://api.aragora.ai", api_key="...")
        >>> result = client.email_priority.prioritize(email)
        >>> ranked = client.email_priority.rank_inbox(emails)
    """

    def __init__(self, client: AragoraClient):
        self._client = client

    # =========================================================================
    # Priority Scoring
    # =========================================================================

    def prioritize(
        self,
        email: dict[str, Any],
        force_tier: ScoringTier | None = None,
    ) -> dict[str, Any]:
        """
        Score a single email for priority.

        Args:
            email: Email message dict with subject, from_address, body, etc.
            force_tier: Force specific scoring tier

        Returns:
            Dict with priority result including score, confidence, rationale
        """
        data: dict[str, Any] = {"email": email}
        if force_tier:
            data["force_tier"] = force_tier
        return self._client.request("POST", "/api/v1/email/prioritize", json=data)

    def rank_inbox(
        self,
        emails: list[dict[str, Any]],
        limit: int | None = None,
    ) -> dict[str, Any]:
        """
        Rank multiple emails by priority.

        Args:
            emails: List of email message dicts
            limit: Maximum results to return

        Returns:
            Dict with ranked results and total count
        """
        data: dict[str, Any] = {"emails": emails}
        if limit:
            data["limit"] = limit
        return self._client.request("POST", "/api/v1/email/rank-inbox", json=data)

    # =========================================================================
    # Feedback and Learning
    # =========================================================================

    def record_feedback(
        self,
        email_id: str,
        action: UserAction,
        email: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """
        Record user action for ML learning.

        Args:
            email_id: Email message ID
            action: User action taken
            email: Optional email data for context

        Returns:
            Dict with success status
        """
        data: dict[str, Any] = {"email_id": email_id, "action": action}
        if email:
            data["email"] = email
        return self._client.request("POST", "/api/v1/email/feedback", json=data)

    def record_feedback_batch(
        self,
        items: list[dict[str, Any]],
    ) -> dict[str, Any]:
        """
        Record batch of user actions.

        Args:
            items: List of feedback items with email_id, action, etc.

        Returns:
            Dict with recorded/errors counts
        """
        return self._client.request("POST", "/api/v1/email/feedback/batch", json={"items": items})

    # =========================================================================
    # Categorization
    # =========================================================================

    def categorize(self, email: dict[str, Any]) -> dict[str, Any]:
        """
        Categorize an email to a smart folder.

        Args:
            email: Email message dict

        Returns:
            Dict with category result and confidence
        """
        return self._client.request("POST", "/api/v1/email/categorize", json={"email": email})

    def categorize_batch(
        self,
        emails: list[dict[str, Any]],
        concurrency: int | None = None,
    ) -> dict[str, Any]:
        """
        Categorize multiple emails.

        Args:
            emails: List of email message dicts
            concurrency: Parallel processing level

        Returns:
            Dict with results and category stats
        """
        data: dict[str, Any] = {"emails": emails}
        if concurrency:
            data["concurrency"] = concurrency
        return self._client.request("POST", "/api/v1/email/categorize/batch", json=data)

    def apply_label(
        self,
        email_id: str,
        category: EmailCategory,
    ) -> dict[str, Any]:
        """
        Apply Gmail label based on category.

        Args:
            email_id: Email message ID
            category: Category to apply

        Returns:
            Dict with success status and applied label
        """
        return self._client.request(
            "POST",
            "/api/v1/email/categorize/apply-label",
            json={"email_id": email_id, "category": category},
        )

    # =========================================================================
    # Inbox Management
    # =========================================================================

    def fetch_inbox(
        self,
        limit: int | None = None,
        offset: int | None = None,
        labels: list[str] | None = None,
        unread_only: bool | None = None,
    ) -> dict[str, Any]:
        """
        Fetch and rank inbox emails.

        Args:
            limit: Maximum emails to fetch
            offset: Number to skip
            labels: Filter by Gmail labels
            unread_only: Only unread emails

        Returns:
            Dict with ranked inbox items
        """
        params: dict[str, Any] = {}
        if limit:
            params["limit"] = limit
        if offset:
            params["offset"] = offset
        if labels:
            params["labels"] = labels
        if unread_only is not None:
            params["unread_only"] = unread_only
        return self._client.request("GET", "/api/v1/email/inbox", params=params if params else None)

    # =========================================================================
    # Configuration
    # =========================================================================

    def get_config(self) -> dict[str, Any]:
        """
        Get email prioritization configuration.

        Returns:
            Dict with VIP lists, thresholds, signal settings
        """
        return self._client.request("GET", "/api/v1/email/config")

    def update_config(self, updates: dict[str, Any]) -> dict[str, Any]:
        """
        Update email prioritization configuration.

        Args:
            updates: Configuration updates

        Returns:
            Dict with updated config
        """
        return self._client.request("PUT", "/api/v1/email/config", json=updates)

    # =========================================================================
    # VIP Management
    # =========================================================================

    def add_vip(
        self,
        email: str | None = None,
        domain: str | None = None,
    ) -> dict[str, Any]:
        """
        Add a VIP email address or domain.

        Args:
            email: VIP email address
            domain: VIP domain

        Returns:
            Dict with success status
        """
        data: dict[str, Any] = {}
        if email:
            data["email"] = email
        if domain:
            data["domain"] = domain
        return self._client.request("POST", "/api/v1/email/vip", json=data)

    def remove_vip(
        self,
        email: str | None = None,
        domain: str | None = None,
    ) -> dict[str, Any]:
        """
        Remove a VIP email address or domain.

        Args:
            email: VIP email to remove
            domain: VIP domain to remove

        Returns:
            Dict with success status
        """
        data: dict[str, Any] = {}
        if email:
            data["email"] = email
        if domain:
            data["domain"] = domain
        return self._client.request("DELETE", "/api/v1/email/vip", json=data)

    # =========================================================================
    # Gmail OAuth
    # =========================================================================

    def get_gmail_oauth_url(
        self,
        redirect_uri: str,
        state: str | None = None,
        scopes: GmailScopes | None = None,
    ) -> dict[str, Any]:
        """
        Get Gmail OAuth authorization URL.

        Args:
            redirect_uri: OAuth callback URL
            state: Optional state parameter
            scopes: Gmail permission scopes

        Returns:
            Dict with oauth_url and scopes
        """
        data: dict[str, Any] = {"redirect_uri": redirect_uri}
        if state:
            data["state"] = state
        if scopes:
            data["scopes"] = scopes
        return self._client.request("POST", "/api/v1/email/gmail/oauth/url", json=data)

    def handle_gmail_oauth_callback(
        self,
        code: str,
        redirect_uri: str,
    ) -> dict[str, Any]:
        """
        Handle Gmail OAuth callback.

        Args:
            code: Authorization code
            redirect_uri: Original redirect URI

        Returns:
            Dict with success status and email
        """
        return self._client.request(
            "POST",
            "/api/v1/email/gmail/oauth/callback",
            json={"code": code, "redirect_uri": redirect_uri},
        )

    def get_gmail_status(self) -> dict[str, Any]:
        """
        Check Gmail connection status.

        Returns:
            Dict with authentication status, email, scopes
        """
        return self._client.request("GET", "/api/v1/email/gmail/status")

    # =========================================================================
    # Cross-Channel Context
    # =========================================================================

    def get_context(self, email_address: str) -> dict[str, Any]:
        """
        Get cross-channel context for an email address.

        Args:
            email_address: Email to look up

        Returns:
            Dict with Slack, calendar, drive activity
        """
        return self._client.request("GET", f"/api/v1/email/context/{email_address}")

    def get_context_boost(self, email: dict[str, Any]) -> dict[str, Any]:
        """
        Get context-based priority boosts for an email.

        Args:
            email: Email message dict

        Returns:
            Dict with boost factors from various signals
        """
        return self._client.request("POST", "/api/v1/email/context/boost", json={"email": email})


class AsyncEmailPriorityAPI:
    """
    Asynchronous Email Priority API.

    Example:
        >>> async with AragoraAsyncClient(base_url="https://api.aragora.ai") as client:
        ...     result = await client.email_priority.prioritize(email)
        ...     await client.email_priority.record_feedback(email_id, "replied")
    """

    def __init__(self, client: AragoraAsyncClient):
        self._client = client

    async def prioritize(
        self,
        email: dict[str, Any],
        force_tier: ScoringTier | None = None,
    ) -> dict[str, Any]:
        """Score a single email for priority."""
        data: dict[str, Any] = {"email": email}
        if force_tier:
            data["force_tier"] = force_tier
        return await self._client.request("POST", "/api/v1/email/prioritize", json=data)

    async def rank_inbox(
        self,
        emails: list[dict[str, Any]],
        limit: int | None = None,
    ) -> dict[str, Any]:
        """Rank multiple emails by priority."""
        data: dict[str, Any] = {"emails": emails}
        if limit:
            data["limit"] = limit
        return await self._client.request("POST", "/api/v1/email/rank-inbox", json=data)

    async def record_feedback(
        self,
        email_id: str,
        action: UserAction,
        email: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Record user action for ML learning."""
        data: dict[str, Any] = {"email_id": email_id, "action": action}
        if email:
            data["email"] = email
        return await self._client.request("POST", "/api/v1/email/feedback", json=data)

    async def record_feedback_batch(
        self,
        items: list[dict[str, Any]],
    ) -> dict[str, Any]:
        """Record batch of user actions."""
        return await self._client.request(
            "POST", "/api/v1/email/feedback/batch", json={"items": items}
        )

    async def categorize(self, email: dict[str, Any]) -> dict[str, Any]:
        """Categorize an email to a smart folder."""
        return await self._client.request("POST", "/api/v1/email/categorize", json={"email": email})

    async def categorize_batch(
        self,
        emails: list[dict[str, Any]],
        concurrency: int | None = None,
    ) -> dict[str, Any]:
        """Categorize multiple emails."""
        data: dict[str, Any] = {"emails": emails}
        if concurrency:
            data["concurrency"] = concurrency
        return await self._client.request("POST", "/api/v1/email/categorize/batch", json=data)

    async def apply_label(
        self,
        email_id: str,
        category: EmailCategory,
    ) -> dict[str, Any]:
        """Apply Gmail label based on category."""
        return await self._client.request(
            "POST",
            "/api/v1/email/categorize/apply-label",
            json={"email_id": email_id, "category": category},
        )

    async def fetch_inbox(
        self,
        limit: int | None = None,
        offset: int | None = None,
        labels: list[str] | None = None,
        unread_only: bool | None = None,
    ) -> dict[str, Any]:
        """Fetch and rank inbox emails."""
        params: dict[str, Any] = {}
        if limit:
            params["limit"] = limit
        if offset:
            params["offset"] = offset
        if labels:
            params["labels"] = labels
        if unread_only is not None:
            params["unread_only"] = unread_only
        return await self._client.request(
            "GET", "/api/v1/email/inbox", params=params if params else None
        )

    async def get_config(self) -> dict[str, Any]:
        """Get email prioritization configuration."""
        return await self._client.request("GET", "/api/v1/email/config")

    async def update_config(self, updates: dict[str, Any]) -> dict[str, Any]:
        """Update email prioritization configuration."""
        return await self._client.request("PUT", "/api/v1/email/config", json=updates)

    async def add_vip(
        self,
        email: str | None = None,
        domain: str | None = None,
    ) -> dict[str, Any]:
        """Add a VIP email address or domain."""
        data: dict[str, Any] = {}
        if email:
            data["email"] = email
        if domain:
            data["domain"] = domain
        return await self._client.request("POST", "/api/v1/email/vip", json=data)

    async def remove_vip(
        self,
        email: str | None = None,
        domain: str | None = None,
    ) -> dict[str, Any]:
        """Remove a VIP email address or domain."""
        data: dict[str, Any] = {}
        if email:
            data["email"] = email
        if domain:
            data["domain"] = domain
        return await self._client.request("DELETE", "/api/v1/email/vip", json=data)

    async def get_gmail_oauth_url(
        self,
        redirect_uri: str,
        state: str | None = None,
        scopes: GmailScopes | None = None,
    ) -> dict[str, Any]:
        """Get Gmail OAuth authorization URL."""
        data: dict[str, Any] = {"redirect_uri": redirect_uri}
        if state:
            data["state"] = state
        if scopes:
            data["scopes"] = scopes
        return await self._client.request("POST", "/api/v1/email/gmail/oauth/url", json=data)

    async def handle_gmail_oauth_callback(
        self,
        code: str,
        redirect_uri: str,
    ) -> dict[str, Any]:
        """Handle Gmail OAuth callback."""
        return await self._client.request(
            "POST",
            "/api/v1/email/gmail/oauth/callback",
            json={"code": code, "redirect_uri": redirect_uri},
        )

    async def get_gmail_status(self) -> dict[str, Any]:
        """Check Gmail connection status."""
        return await self._client.request("GET", "/api/v1/email/gmail/status")

    async def get_context(self, email_address: str) -> dict[str, Any]:
        """Get cross-channel context for an email address."""
        return await self._client.request("GET", f"/api/v1/email/context/{email_address}")

    async def get_context_boost(self, email: dict[str, Any]) -> dict[str, Any]:
        """Get context-based priority boosts for an email."""
        return await self._client.request(
            "POST", "/api/v1/email/context/boost", json={"email": email}
        )
